### PR TITLE
[IMP] purchase,sale: Take into account order date for sequence

### PR DESCRIPTION
**Pull request by pedrobaeza**
Version of #28921 for upcoming v13

**Steps to reproduce**

* Define "Sales Order" sequence for using date ranges checking "Use subsequences per date_range".
* Change sequence prefix to "SO/%(year)s/".
* Create a sales order with order date in 2019 (a different year than current one).

**Current behavior**

Got a sales order with 2018 (current year) number.

**Expected behavior**

Got a sales order with 2019 number.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

cc @Tecnativa

### CODE REVIEW

**Comment by dbo-odoo** [comment]
@pedrobaeza when you manage to get it to work (you can add a test if you want :p), can you ping me and @sle-odoo (he's in charge of purchase)? You can check also tomorrow after him with him (or someone from the logistics team)

**Comment by pedrobaeza** [comment]
@dbo-odoo the problem is in `ir.sequence`, that it's not using proper date for string conversion (`%(y)s`), but the sequence range is being properly searched and used for next number. I think it's a leftover when `sequence_date` argument was introduced replacing context argument. We have 2 options:

- To fix `ir.sequence` for supporting properly the argument. This may need maybe to change any signature, but it should be the way to go IMO.
- Use context key here instead.

What do you want me to do?

File: addons/purchase/models/purchase.py [hunk 1]
```diff
@@ -156,7 +156,10 @@ def name_get(self):
     @api.model
     def create(self, vals):
         if vals.get('name', 'New') == 'New':
-            vals['name'] = self.env['ir.sequence'].next_by_code('purchase.order') or '/'
+            seq_date = None
+            if 'date_order' in vals:
+                seq_date = fields.Date.to_string(fields.Date.to_date(vals['date_order']))
+            vals['name'] = self.env['ir.sequence'].next_by_code('purchase.order', sequence_date=seq_date) or '/'
```
Comment (sle-odoo) [inline addons/purchase/models/purchase.py]: seems like `sequence_date` expectes a Date object, i don't think you need the outer ```fields.Date.to_string``` ?

**Comment by sle-odoo** [comment]
asides from my remark (which i'm not sure) the patch looks good, but i would like you to write a test (purchase is dramatically in needs of new testcases :) )

**Comment by dbo-odoo** [comment]
@pedrobaeza there is no issue with sequences, you just have to use `%(range_year)s` instead of `%(year)s` (read the little helper text at the bottom of the sequence's form view). Don't feel bad, I missed it as well :p
[Demo](https://www.youtube.com/watch?v=Eudoovtm_70&feature=youtu.be)

**Comment by pedrobaeza** [comment]
Ouch, that hurts, how can I missed it! OK then, I'll prepare the tests and tell you back.

What I miss then is the possibility to put the "range year" with only 2 digits, but that's another history...

File: addons/purchase/models/purchase.py (hunk 1 above)
Comment (dbo-odoo) [reply addons/purchase/models/purchase.py]: Agreed, i don't understand the double conversion. The `to_date` call should be enough. Same for `sale`.

**Comment by dbo-odoo** [comment]
`%(range_y)s` works fine, check the video :)

**Comment by pedrobaeza** [comment]
OK, I only read the message. Perfect then. Doing the tests now.

**Comment by pedrobaeza** [comment]
OK, everything fixed and tests added. I have retargeted to 13.0, but not sure if this still needs to be kept on saas-12.5.

Thanks to all for your advices.

**Comment by dbo-odoo** [comment]
One last issue encountered during testing: this code does not take timezone difference into account.

How to reproduce:
1. Configure your purchase sequence with range, e.g. from Jan 1 to Dec 31 for 2019 and 2020 and with an extrapolation of `P/%(range_year)s/`
1. Ensure you are ahead of the UTC timezone (should be the case for Spain by default)
1. Create a PO with order date = `01/01/2020 00:00:00` (in the web client, which will convert to UTC itself)
=> the number it gets is `P/2019/...` which is not correct

Do the same test when you change your order date to e.g. `01/01/2020 03:00:00` (assuming you are less than 3 hours ahead of UTC) => the range year in the numbering is correct

I think this should do the trick:
``` python
seq_date = fields.Datetime.context_timestamp(self, fields.Datetime.to_datetime(vals['date_order']))
```

The test can be made to fail for this particular case with something like this:
```python
# In EU/BXL tz, this is actually already 01/01/2020
vals['date_order'] = '2019-12-31 23:30:00'
purchase_order = PurchaseOrder.with_context(tz='Europe/Brussels').create(vals.copy())
self.assertTrue(purchase_order.name.startswith('PO/2020/'))
```

The problem is that timezone conversion of data coming from the client is done in the client itself, so even by using midnight fringe cases in the values, this is invisible during testing.

Note: `context_timestamp` first arg is named `record`, but it just uses it to get the `tz` from the env bound to the record, so an empty recordset works as well.

**Comment by pedrobaeza** [comment]
@dbo-odoo I have pushed your latest comments applied to both sale and purchase. Thanks for the advices.

**Comment by dbo-odoo** [comment]
@robodoo r+

**Comment by dbo-odoo** [comment]
@robodoo rebase-ff

**Pull request closed without merging**